File: dotnet-soa/ProductService/Infrastructure/Web/Dto/ProductResponse.cs
namespace ProductService.Infrastructure.Web.Dto;

/// <summary>
/// DTO for product API responses.
/// Decouples the API response format from the internal domain model.
/// </summary>
public record ProductResponse
{
    public Guid Id { get; init; }
    public string Name { get; init; } = string.Empty;
    public string? Description { get; init; }
    public decimal Price { get; init; }
    public string? Category { get; init; }
    public DateTime CreatedAt { get; init; }
    public DateTime UpdatedAt { get; init; }
}

File: java-mvc/hexagonal-architecture/src/main/java/com/ecommerce/hexagonal/adapter/input/web/dto/ProductResponse.java
package com.ecommerce.hexagonal.adapter.input.web.dto;

import com.ecommerce.hexagonal.domain.model.Product;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

public record ProductResponse(
        UUID id,
        String name,
        String description,
        BigDecimal price,
        String category,
        Integer stockQuantity,
        String sku,
        boolean active,
        boolean inStock,
        LocalDateTime createdAt,
        LocalDateTime updatedAt
) {
    public static ProductResponse fromDomain(Product product) {
        return new ProductResponse(
                product.getId(),
                product.getName(),
                product.getDescription(),
                product.getPrice(),
                product.getCategory(),
                product.getStockQuantity(),
                product.getSku(),
                product.isActive(),
                product.isInStock(),
                product.getCreatedAt(),
                product.getUpdatedAt()
        );
    }
}
